In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import tensorflow as tf
from tensorflow import keras

In [2]:
import warnings
warnings.filterwarnings('ignore')

df=pd.read_csv('sonar_dataset.csv',header=None)
df.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      51      52      53      54      55      56      57  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   

       58      59  60  
0  0.0090  0.0032   R  
1  0.0052  0.0044   R  
2  0.0095  0.0078   R  
3  0.0040  0.0117   R  
4  0.0107  0.0094   R  

[5 rows x 61 columns]

In [3]:
df.dtypes

0     float64
1     float64
2     float64
3     float64
4     float64
       ...   
56    float64
57    float64
58    float64
59    float64
60     object
Length: 61, dtype: object

In [4]:
df.shape

(208, 61)

In [5]:
df.isna().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [6]:
df.columns

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
            51, 52, 53, 54, 55, 56, 57, 58, 59, 60],
           dtype='int64')

In [7]:
df[60].value_counts()

M    111
R     97
Name: 60, dtype: int64

111 samples of metal cylinder & 97 samples of rock

Now dividing the data

In [8]:
x=df.drop(60,axis=1)
y=df[60]

Changing y object type to numeric by using one hot encoding

In [9]:
y=pd.get_dummies(y,drop_first=True)
y.head(5)
#R--1 & M--0

R
0  1
1  1
2  1
3  1
4  1

In [15]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,random_state=1,test_size=0.25)


In [17]:
x_train.shape,x_test.shape

((156, 60), (52, 60))

In [19]:
model=keras.Sequential([
    keras.layers.Dense(60,input_dim=60, activation='relu'),
    keras.layers.Dense(30,activation='relu'),
    keras.layers.Dense(15,activation='relu'),
    keras.layers.Dense(1,activation='sigmoid')
])

model.compile(loss='binary_crossentropy',optimizer='adam',metrics=['accuracy'])

model.fit(x_train,y_train,epochs=100,batch_size=8)

Epoch 1/100
20/20 [==============================] - 0s 770us/step - loss: 0.7061 - accuracy: 0.4872
Epoch 2/100
20/20 [==============================] - 0s 777us/step - loss: 0.6817 - accuracy: 0.5385
Epoch 3/100
20/20 [==============================] - 0s 733us/step - loss: 0.6723 - accuracy: 0.5192
Epoch 4/100
20/20 [==============================] - 0s 724us/step - loss: 0.6548 - accuracy: 0.5705
Epoch 5/100
20/20 [==============================] - 0s 702us/step - loss: 0.6344 - accuracy: 0.6474
Epoch 6/100
20/20 [==============================] - 0s 755us/step - loss: 0.6052 - accuracy: 0.6987
Epoch 7/100
20/20 [==============================] - 0s 690us/step - loss: 0.5765 - accuracy: 0.7308
Epoch 8/100
20/20 [==============================] - 0s 701us/step - loss: 0.5453 - accuracy: 0.8013
Epoch 9/100
20/20 [==============================] - 0s 735us/step - loss: 0.5149 - accuracy: 0.7692
Epoch 10/100
20/20 [==============================] - 0s 734us/step - loss: 0.4863 - accura

We can see model is overfit

In [20]:
model.evaluate(x_test,y_test)

2/2 [==============================] - 0s 999us/step - loss: 1.0707 - accuracy: 0.7885


[1.0707484483718872, 0.7884615659713745]

In [25]:
y_pred=model.predict(x_test).reshape(-1)
y_pred=np.round(y_pred)

In [28]:
y_pred[:10]

array([0., 1., 1., 0., 1., 1., 0., 1., 0., 1.], dtype=float32)

In [29]:
y_test[:10]

R
186  0
155  0
165  0
200  0
58   1
34   1
151  0
18   1
202  0
62   1

In [31]:
from sklearn.metrics import confusion_matrix, classification_report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.75      0.89      0.81        27
           1       0.85      0.68      0.76        25

    accuracy                           0.79        52
   macro avg       0.80      0.78      0.78        52
weighted avg       0.80      0.79      0.79        52



## Dropout layer

In [36]:
model1=keras.Sequential([
    keras.layers.Dense(60,input_dim=60, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(30,activation='relu'),
    keras.layers.Dense(15,activation='relu'),
    keras.layers.Dense(1,activation='sigmoid')
])

model1.compile(loss='binary_crossentropy',optimizer='adam',metrics=['accuracy'])

model1.fit(x_train,y_train,epochs=100,batch_size=8)

Epoch 1/100
20/20 [==============================] - 0s 807us/step - loss: 0.7107 - accuracy: 0.4551
Epoch 2/100
20/20 [==============================] - 0s 697us/step - loss: 0.6860 - accuracy: 0.5833
Epoch 3/100
20/20 [==============================] - 0s 693us/step - loss: 0.6885 - accuracy: 0.5513
Epoch 4/100
20/20 [==============================] - 0s 719us/step - loss: 0.6730 - accuracy: 0.5897
Epoch 5/100
20/20 [==============================] - 0s 696us/step - loss: 0.6753 - accuracy: 0.5833
Epoch 6/100
20/20 [==============================] - 0s 742us/step - loss: 0.6543 - accuracy: 0.6282
Epoch 7/100
20/20 [==============================] - 0s 739us/step - loss: 0.6477 - accuracy: 0.6090
Epoch 8/100
20/20 [==============================] - 0s 772us/step - loss: 0.6664 - accuracy: 0.6154
Epoch 9/100
20/20 [==============================] - 0s 783us/step - loss: 0.6116 - accuracy: 0.7179
Epoch 10/100
20/20 [==============================] - 0s 738us/step - loss: 0.6343 - accura

In [37]:
model1.evaluate(x_test,y_test)

2/2 [==============================] - 0s 1ms/step - loss: 0.4650 - accuracy: 0.7885


[0.4649783670902252, 0.7884615659713745]

In [39]:
y_pred=model1.predict(x_test).reshape(-1)
y_pred=np.round(y_pred)

In [40]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.75      0.89      0.81        27
           1       0.85      0.68      0.76        25

    accuracy                           0.79        52
   macro avg       0.80      0.78      0.78        52
weighted avg       0.80      0.79      0.79        52

